### PR TITLE
CA-353388: [XSI-989] stunnel sub-process logs fillup partition

### PR DESCRIPTION
**Pull request by liulinC**
The log level is set to debug, which involve lots of logs fillup partition
This commit reset the log level to warning

Signed-off-by: Lin Liu <lin.liu@citrix.com>

### CODE REVIEW

File: stunnel/stunnel.ml
Comment (lippirk) [inline stunnel/stunnel.ml:193]: Do we not want level 7 when `extended_diagnosis = true`? I'd have thought we should log as much as possible in this case

Comment (liulinC) [reply stunnel/stunnel.ml:193]: The extended_diagnosis  is enabled by default, and it will fill up disk partition.
I cannot fix it by just  disable the `extended_diagnosis` as it do more things than logging.

Comment (lindig) [reply stunnel/stunnel.ml:193]: I believe we don't need the extended logs. However, log rotation should avoid the problem. I'm fine with reducing the log level for Stunnel.

Comment (liulinC) [reply stunnel/stunnel.ml:193]: btw,  the main process just hardcode to 5, see
https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/helpers.ml#L1739
Do we need to hard code 5 here to keep consistency?

Comment (lippirk) [reply stunnel/stunnel.ml:193]: > I cannot fix it by just disable the extended_diagnosis as it do more things than logging

What else is it used for?

> The extended_diagnosis is enabled by default

I think changing the default makes more sense. Could we change this line instead perhaps? https://github.com/xapi-project/xen-api-libs-transitional/blob/c77ad87cb6f34894f32ccf051e3bf0bc7c12ff45/http-svr/xmlrpc_client.ml#L327

We use extended_diagnosis=true for the slave -> master connection. It might be useful to keep verbose logging here

Comment (liulinC) [reply stunnel/stunnel.ml:193]: extended_diagnosis controls not only the log level, but also something like whether remove some log files,  we may not want to just disable it here, 

also, I do not find a good reason to enable debug log by default.

Comment (lippirk) [reply stunnel/stunnel.ml:193]: > extended_diagnosis controls not only the log level, but also something like whether remove some log files, we may not want to just disable it here,

where does it cause log files to be removed? I don't see anything related to that in the usages? https://github.com/xapi-project/xen-api-libs-transitional/search?q=extended_diagnosis

Comment (lippirk) [reply stunnel/stunnel.ml:193]: Ah, I see it now (github search completely missed it..) https://github.com/xapi-project/xen-api-libs-transitional/blob/c77ad87cb6f34894f32ccf051e3bf0bc7c12ff45/stunnel/stunnel.ml#L316

Comment (lippirk) [reply stunnel/stunnel.ml:193]: I think it'd be good to just not delete that file ever, and change the default. But I think the current solution is also fine

**Review by lippirk — APPROVED** [review]

File: stunnel/stunnel.ml
Comment (liulinC) [reply stunnel/stunnel.ml:193]: @lindig  please help to approve this PR if you are happy with it, thanks.

**Review by lindig — APPROVED** [review]

**Comment by lindig** [comment]
I think a weakness of the whole design is that we can't change the logging level in a running installation even when restarting the daemon. A better design could watch  for some environment variable that we could set to enable more logging. Or the function could take an optional parameter with a level that we supply at the call site. This might be more effort than it is worth, though.

**Comment by liulinC** [comment]
@lindig @lippirk @minglumlu  I have refined the solution which can be possible enable debug by `debug_stunnel` environment variable,  
Also, align with xapi, https://github.com/xapi-project/xen-api/pull/4382

Could you please help to check again whether you are happy with it, thanks in advance.

File: stunnel/stunnel.ml
Comment (lippirk) [inline stunnel/stunnel.ml:167]: I'd prefer if callers didn't have to rely on environment variables. We could expose two functions:

```suggestion
let debug_conf_of_bool verbose : string = if verbose then "debug=authpriv.7" else "debug=authpriv.5"

let debug_conf_of_env () : string =
    (try Unix.getenv "debug_stunnel" with _ -> "") |> 
    String.lowercase_ascii |> 
    fun x -> List.mem x ["yes"; "true"; "1"] |>
    debug_conf_of_bool
```

Please can you also document `extended_diagnosis`, as it's not obvious what the purpose of that is now

**Comment by liulinC** [comment]
Thanks @lippirk @lindig  for the help,  
I have updated the. PR basing on the suggestions, could you please help to check again.
BTW, I will squash the commits before merge, just keep it for now to make it easier for the reviewers.

**Review by lippirk — APPROVED** [review]

**Comment by minglumlu** [comment]
I just wonder how to use this env variable to enable verbose debug log, especially for the client connections initiated by XAPI?

**Review by minglumlu — APPROVED** [review]

**Pull request merged**
